src/core/main/Reflection/InstanceAnalyzer.cs

```diff
@@ -0,0 +1,274 @@
+            
+            var typeInfo = instance.GetType().GetTypeInfo();
+
+            //
+            // fields
+            //
+            var fields = typeInfo.GetFields(bindingFlags);
+            foreach (var fieldInfo in fields)
+            {
+                if (!IsBackingField(fieldInfo))
+                {
+                    FieldAndPropertyHandler(listener, fieldInfo, context, instance);
+                }
+            }
+            
+            //
+            // properties
+            //
+            var props = typeInfo.GetProperties(bindingFlags);
+            foreach (var propertyInfo in props)
+            {
+                FieldAndPropertyHandler(listener, propertyInfo, context, instance);
+            }
+            
+            //
+            // methods
+            //
+            var methods = typeInfo.GetMethods(bindingFlags);
+            foreach (var methodInfo in methods)
+            {
+                if (
+                    methodInfo.DeclaringType != typeof(System.Object) // do not "announce" methods like Object.ToString
+                    && !IsBackingMethod(methodInfo) // do not "announce" auto property backing methods
+                )
+                {
+                    listener.OnMethod(methodInfo, context);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Call OnField or OnProperty for the given member
+        /// </summary>
+        private void CallListener(IInstanceListener listener, InstanceAnalyzerContext context, MemberInfo memberInfo, Func<object> valueGetter)
+        {
+            if (memberInfo.MemberType == MemberTypes.Field)
+            {
+                listener.OnField((FieldInfo) memberInfo, valueGetter, context);
+            }
+            else
+            {
+                listener.OnProperty((PropertyInfo) memberInfo, valueGetter, context);
+            }
+        }
+
+        /// <summary>
+        /// Field and properties are actually quite similar
+        /// in this context. This method contains "generic" handling
+        /// of both types, so we avoid duplication. 
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Thrown if we somehow try to iterate something that is not an IEnumerable</exception>
+        private void FieldAndPropertyHandler(IInstanceListener listener, MemberInfo memberInfo, InstanceAnalyzerContext context, object instance)
+        {
+            Func<object> valueGetter = () => memberInfo.GetValue(instance);
+            
+            CallListener(listener, context, memberInfo, valueGetter);
+            
+            var valueType = memberInfo.GetTypeOfValue();
+
+            if (ShouldIterate(valueType))
+            {
+                object value = valueGetter.Invoke();
+                
+                if (value != null)
+                {
+                    if (value is IDictionary)
```

Comment:
So it supports `IDictionary` and `IEnumerable`. Are there other "exotic" list types that we should consider supporting or does `IEnumerable` and `IDictionary` cover'em'all?

Reply:
As far as I can see everything in the `System.Collections` namespace has `IDictionary` and/or `IEnumerable`.